File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::unordered_map<char, char> brackets = {
            {'(', ')'}, 
            {'{', '}'}, 
            {'[', ']'}
        };
        std::stack<char> st;

        for (char c: s) {
            if (brackets.count(c)) {
                st.push(c);
            } else {
                if (st.empty() || brackets[st.top()] != c) {
                    return false;
                } 
                st.pop();
            }
        }

        return st.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        brackets = {'(' : ')', '{': '}', '[': ']'}

        for c in s:
            if c in brackets:
                stack.append(c)
            else:
                if not stack or brackets[stack[-1]] != c:
                    return False
                stack.pop()
                
        return len(stack) == 0